plugins/maven-dependency-resolver/src/main/java/org/robolectric/internal/dependency/MavenArtifactFetcher.java

```diff
@@ -0,0 +1,198 @@
+          .callAsync(
+              () -> {
+                // double check that the artifact has not been installed
+                if (new File(localRepositoryDir, artifact.jarPath()).exists()) {
+                  removeArtifactFiles(stagingRepositoryDir, artifact);
+                  return Futures.immediateFuture(null);
+                }
+                createArtifactSubdirectory(artifact, localRepositoryDir);
+                boolean pomValid = validateStagedFiles(artifact.pomPath(), artifact.pomSha1Path());
+                if (!pomValid) {
+                  throw new AssertionError("Unable to validate POM file");
+                }
+                boolean jarValid = validateStagedFiles(artifact.jarPath(), artifact.jarSha1Path());
+                if (!jarValid) {
+                  throw new AssertionError("Unable to validate JAR file");
+                }
+                Logger.info(
+                    String.format(
+                        "Checksums validated, moving artifact %s to local maven directory",
+                        artifact));
+                commitFromStaging(artifact.pomPath());
+                commitFromStaging(artifact.jarPath());
+                commitFromStaging(artifact.jarSha1Path());
+                commitFromStaging(artifact.pomSha1Path());
+                removeArtifactFiles(stagingRepositoryDir, artifact);
+                return Futures.immediateFuture(null);
+              },
+              executorService)
+          .get();
+    } catch (InterruptedException | ExecutionException e) {
+      if (e instanceof InterruptedException) {
+        Thread.currentThread().interrupt(); // Restore the interrupted status
+      }
+      removeArtifactFiles(stagingRepositoryDir, artifact);
+      removeArtifactFiles(localRepositoryDir, artifact);
+      Logger.error("Failed to fetch maven artifact " + artifact, e);
+      throw new AssertionError("Failed to fetch maven artifact " + artifact, e);
+    }
+  }
+
+  private void removeArtifactFiles(File repositoryDir, MavenJarArtifact artifact) {
+    new File(repositoryDir, artifact.jarPath()).delete();
+    new File(repositoryDir, artifact.jarSha1Path()).delete();
+    new File(repositoryDir, artifact.pomPath()).delete();
+    new File(repositoryDir, artifact.pomSha1Path()).delete();
+  }
+
+  private boolean validateStagedFiles(String filePath, String sha1Path) throws IOException {
+    File tempFile = new File(this.stagingRepositoryDir, filePath);
+    File sha1File = new File(this.stagingRepositoryDir, sha1Path);
+
+    HashCode expected = HashCode.fromString(new String(Files.asByteSource(sha1File).read(), UTF_8));
+
+    HashCode actual = Files.asByteSource(tempFile).hash(Hashing.sha1());
+    return expected.equals(actual);
+  }
+
+  private void createArtifactSubdirectory(MavenJarArtifact artifact, File repositoryDir) {
+    File artifactDirectory = new File(repositoryDir, artifact.jarPath()).getParentFile();
+    artifactDirectory.mkdirs();
```

Comment:
Nit: You should check the return value of `mkdirs` and throw eventually.

Reply:
Done